README.md
# matebook-applet
System tray applet/control app for Huawei Matebook

![Build Status](https://github.com/nekr0z/matebook-applet/actions/workflows/build.yml/badge.svg) [![codecov](https://codecov.io/gh/nekr0z/matebook-applet/branch/master/graph/badge.svg)](https://codecov.io/gh/nekr0z/matebook-applet)

---

## WARNING

Due to [a recent Codecov security problem](https://about.codecov.io/security-update/) the PGP key that was used to sign binaries and the Debian repository could have been compromised. It is strongly recommended to remove the old key from your trusted `apt` keyring since it can no longer really be trusted:
```
sudo apt-key del F25E85CB21A79726
```
The repository uses a new key now, you can add it [by usual means](#debian).

---

This simple applet is designed to make some of the proprietary Huawei PC Manager's functionality available via GUI on Linux.
 
![Applet screenshot](screenshot-applet.png?raw=true "matebook-applet")

It can be used as a system tray applet (above) or as a windowed app (below).

![Windowed screenshot](screenshot-windowed.png?raw=true "matebook-applet windowed mode")

##### Table of Contents
* [Installation and setup](#installation-and-setup)
  * [Debian and derivatives](#debian)
  * [Other Linux](#other-linux) - kernel 5.0 and newer
  * [Old Linux](#old-linux) - for those with pre-5.0 kernel
  * [Compiling](#compiling-matebook-applet) by yourself
* [Usage](#usage)
  * [Gnome](#gnome)
* [Development](#development)
  * [Contributing translations](#contributing-translations)
* [Credits](#credits)

#### Help `matebook-applet` get better!
Join the [development](#development)! You can also support by sending BTC to `3N983ChqZyFtiep6tNu4tnNcQnDtY2QWps`, or to Pay ID `435921757` on Binance.

## Installation and setup
The applet requires no installation as such. However, it has dependencies, its functionality may be limited without proper setup, and you may want to add it to autorun so that it gets started automatically when you load your DE. Read on for the details.

### Debian
Debian users can add this repo:
```
deb [signed-by=/usr/share/keyrings/matebook-applet.key] http://evgenykuznetsov.org/repo/ stable main
```
to their `/etc/apt/sources.list` and add the public key:
```
$ wget -qO - https://raw.githubusercontent.com/nekr0z/matebook-applet/master/matebook-applet.key | sudo tee /usr/share/keyrings/matebook-applet.key
```

From there, the applet is just a `sudo apt install matebook-applet` away.

Users report that this way also works on Debian derivatives such as Ubuntu or Linux Mint.

<details>
<summary><b>note to pre-2022 users</b></summary>

The repository used to be signed by other public keys, and this README used to recommend adding the key to the system-wide trusted keyring. This practice is no longer recommended by Debian because the key is then trusted by the system to sign any known repository, which is a security weakness. If you installed matebook-applet `.deb` package via this repository before June 2021, you may still have the old keys in your system-wide trusted keyring, and it's a good idea to remove them:
```
sudo apt-key del 466F4F38E60211B0
sudo apt-key del F25E85CB21A79726
sudo apt-key del FA32B7DDA1A3AC2C
```

</details>

### Other Linux
If you're running a Linux with relatively new kernel version, it already includes the required driver. Make sure your system has GTK+ and `libayatana-appindicator` installed, download (from the [releases page](https://github.com/nekr0z/matebook-applet/releases)) or [build](#compiling-matebook-applet) `matebook-applet`, and it should work *in read-only mode* right away.

If your system doesn't have `libayatana-appindicator`, you can use a legacy build that depends on `libappindicator` instead.

For pre-5.5 kernels you may need to update the [Huawei-WMI driver](https://github.com/aymanbagabas/Huawei-WMI). The applet requres at least version 3.0 of the driver.

To be able *to change settings* as opposed to simply displaying them, you either need to run the applet as root (absolutely not recommended), or make sure all the necessary files (the hooks in `/sys/devices/platform/huawei-wmi` as well as `/etc/default/huawei-wmi/` directory) are user-writable. A good way to set everything up is to make use of [Rouven Spreckels](https://github.com/n3vu0r)' awesome [project](https://github.com/qu1x/huawei-wmi):
```
$ git clone https://github.com/qu1x/huawei-wmi.git
$ cd huawei-wmi/generic
$ sudo make install
```
You may need to re-login for adding your user to group to take effect.

### Old Linux

On Linux kernels earlier than 5.0 the Huawei-WMI driver is not available, so the alternative method should be used.

<details>
<summary><b>details</b></summary>

You can make `matebook-applet` work using the `-r` command line option and two scripts as explained below. Any one of the scripts is enough, but you need both for full functionality. You may download them both in one archive [here](https://github.com/nekr0z/linux-on-huawei-matebook-13-2019/releases). Please note that these scripts are for MateBook 13 and will not work on other MateBooks (you can still get them to work by changing them to address EC registers proper for your particular model, but make sure you know exactly what you're doing, you've been warned).

Both scripts depend on [ioport](https://people.redhat.com/~rjones/ioport/) to work, so make sure to install it first. Debian packages are available in main repository, and there is [a way](https://github.com/nekr0z/matebook-applet/issues/5) to get it on Arch, too.

* To get battery protection functionality:

    1. Download the [batpro script](https://github.com/nekr0z/linux-on-huawei-matebook-13-2019/blob/master/batpro), make it executable and copy to root executables path, i.e.:
        ```
        # mv batpro /usr/sbin/
        ```
    2. Allow your user to execute the script without providing sudo credentials. One way to do this would be to explicitly add sudoers permission (please be aware of possible security implications):
        ```
        # echo "your_username ALL = NOPASSWD: /usr/sbin/batpro" >> /etc/sudoers.d/batpro
        # chmod 0440 /etc/sudoers.d/batpro
        ```
  3. Double check that you can successfully run the script without providing additional authentication (i.e. password):
        ```
        $ sudo batpro status
        ```
* To get Fn-Lock functionality:

    1. Download the [fnlock script](https://github.com/nekr0z/linux-on-huawei-matebook-13-2019/blob/master/fnlock), make it executable and copy to root executables path, i.e.:
        ```
        # mv fnlock /usr/sbin/
        ```
    2. Allow your user to execute the script without providing sudo credentials. One way to do this would be to explicitly add sudoers permission (please be aware of possible security implications):

            # echo "your_username ALL = NOPASSWD: /usr/sbin/fnlock" >> /etc/sudoers.d/fnlock
            # chmod 0440 /etc/sudoers.d/fnlock

    3. Double check that you can successfully run the script without providing additional authentication (i.e. password):

            $ sudo fnlock status

</details>

### Compiling matebook-applet
You can always download precompiled amd64 binary from the [releases page](https://github.com/nekr0z/matebook-applet/releases), but it's also perfectly OK to compile matebook-applet yourself. Provided that you have the dependencies (GTK+ and libayatana-appindicator) installed (on Debian you can `sudo apt install libgtk-3-dev libayatana-appindicator3-dev`), all you need to do is:

    $ git clone https://github.com/nekr0z/matebook-applet.git
    $ cd matebook-applet
    $ go build

To build against `libappindicator` instead, append the last command with `-tags=legacy_appindicator`.

## Usage
The user interface is intentionally as simple as they get. You get an icon in system tray that you can click and get a menu. The menu consists of current status, options to change it, and an option to quit the applet. Please be aware that the applet does not probe for current status on its own (this is intentional), so if you change your battery protection settings by other means it will not reflect the change. Clicking on the status line (top of the menu) updates it.

The entry that shows current Fn-Lock status is clickable, too, that toggles Fn-Lock (from ON to OFF or vice versa). Again, no probing here, so if you change Fn-Lock status by other means it will not reflect the change until clicked, but then it will toggle Fn-Lock again.

Command line option `-w` launches the applet in windowed (app) mode, i.e.:
```
$ matebook-applet -w
```

Other command line options can be found on the included manpage:
```
$ man -l matebook-applet.1
  or, if you installed applet from repository or .deb package,
$ man matebook-applet
```
### Gnome
As of Gnome 3.26 [the "legacy tray" is removed](https://bugzilla.gnome.org/show_bug.cgi?id=785956). Launching the applet in windowed (app) mode still works. An extension is needed for the system tray icon to show up, for example [AppIndicator](https://github.com/ubuntu/gnome-shell-extension-appindicator). 

Also, there is a separate native extension by another developer: [Huawei WMI Controls](https://extensions.gnome.org/extension/4580/huawei-wmi-controls), which provides an interface similar to `matebook-applet`.

## Development
Pull requests are always welcome!

### Contributing translations
> In case you don't want to install any Go tools, the notion of having to use `git` scares you, and you don't feel like figuring out what a pull request is and how to make one, but you still feel like contributing a translation, just drop me an email, we'll figure something out. ;-)

The intended way to go about contributing translations to the project is using [go-i18n](https://github.com/nicksnyder/go-i18n) tool. Make sure you have it installed and available, clone the repository, and:
```
$ cd assets/translations
$ ls
$ touch active.xx.toml
```
(where `xx` is your language)
```
$ goi18n merge active.*.toml
```
This will generate `translate.xx.toml` with all the lines that need to be translated to the language `xx`. Go ahead and translate it. Please make sure you delete all the messages you don't feel like translating before the next step, so that English messages don't creep in disguised as messages in your language.
```
$ goi18n merge active.*.toml translate.*.toml
```
Now all you've translated is in `active.xx.toml` and you can make a pull request.

## Credits
This software includes the following software or parts thereof:
* [getlantern/systray](https://github.com/getlantern/systray) Copyright © 2014 Brave New Software Project, Inc.
* [andlabs/ui](https://github.com/andlabs/ui) Copyright © 2014 Pietro Gagliardi
* [BurntSushi/toml](https://github.com/BurntSushi/toml) Copyright © 2013 TOML authors
* [Jibber Jabber](https://github.com/cloudfoundry/jibber_jabber) Copyright © 2014 Pivotal
* [go-i18n](https://github.com/nicksnyder/go-i18n) Copyright © 2014 [Nick Snyder](https://github.com/nicksnyder)
* [The Go Programming Language](https://golang.org) Copyright © 2009 The Go Authors
* [Enkel Icons](https://www.deviantart.com/froyoshark/art/WIP-Enkel-Icon-Pack-for-Mac-498003378) by [FroyoShark](https://www.deviantart.com/froyoshark)

Translations:
* German - [Rouven "n3vu0r" Spreckles](https://github.com/n3vu0r)
* Russian - [Evgeny "nekr0z" Kuznetsov](https://evgenykuznetsov.org)
* Spanish - [ACV99](https://github.com/ACV99)

Packages are built using [fpm](https://github.com/jordansissel/fpm) and [changelog](https://evgenykuznetsov.org/go/changelog).

Big **THANK YOU** to [Ayman Bagabas](https://github.com/aymanbagabas) for all his work and support. Without him, there would be no matebook-applet.

Kudos to [Rouven Spreckels](https://github.com/n3vu0r) for sorting out `udev` and `systemd`.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [3.1.0] - 2023-05-05
### Added
- keyboard light timeout control (for the driver version that supports it)

### Changed
- update dependencies
- update Go to 1.19

## [3.0.3] - 2022-03-17
### Changed
- update Go to 1.16
- remove `vfsgen` dependency (use standard-library `embed` instead)
- update dependencies
- revert to using upstream `systray` library that now has `libayatana` support
- make build tag consistent with updated `systray` library

## [3.0.2] - 2021-09-03
### Translations
- Spanish

## [3.0.1] - 2021-06-30
### Fixed
- provide legacy builds against `libappindicator`

## [3.0.0] - 2021-06-10
### Changed
- use `libayatana-appindicator` instead of the deprecaded `libappindicator`
- scripts are not attempted by default (i.e. `-r` is false by default)
### Removed
- the long-deprecated `-s` flag

## [2.5.0] - 2020-12-19
### Added
- information on kernel 5.5 in README
### Changed
- BP values of `0 0` returned from the driver are no longer considered invalid but are treated as BP OFF, since this seems to be the case sometimes, especially on 2020 models

## [2.4.8] - 2019-12-11
### Changed
- migrated Debian repo to a new signing key

## [2.4.7] - 2019-11-12
### Translations
- German

## [2.4.6] - 2019-11-12
### Fixed
- signing binaries

## [2.4.5] - 2019-11-12
### Fixed
- releasing debian package

## [2.4.4] - 2019-11-12
### Fixed
- signing binaries

## [2.4.3] - 2019-11-10
### Fixed
- release script

## [2.4.2] - 2019-11-10
### Changed
- prepare to migrate Debian repo to a new signing key

## [2.4.1] - 2019-11-06
### Fixed
- changelog
### Translations
- translated flags usage

## [2.4.0] - 2019-11-06
### Added
- internationalization
### Translations
- Russian

## [2.3.3] - 2019-10-31
### Fixed
- build script

## [2.3.2] - 2019-10-31
### Fixed
- release script

## [2.3.1] - 2019-10-31
### Fixed
- lack of proper changelog
- displayed version numbers

## [2.3.0] - 2019-09-22
### Added
- windowed mode (triggered by `-w` option)
- deprecation warning when `-s` option is used
### Fixed
- crashes in cases where either only fnlock or only battery endpoint is available
- lack of documentation about GTK+ dependency

## [2.2.0] - 2019-09-21
### Added
- a way to set custom battery threshold

## [2.1.4] - 2019-09-19
### Changed
- platform endpoints are now preferred instead of kernel ones
### Fixed
- persistent battery thresholds endpoint autodetection
- minor glitches in `*.deb` generation

## [2.1.3] - 2019-09-16
### Added
- `-r` option for using scripts as endpoints; on by default for now, but will be off by default eventually
### Fixed
- `batpro` script not working
- minor glitches

## [2.1.2] - 2019-09-15
### Fixed
- support for some corner cases and legacy endpoints

## [2.1.1] - 2019-09-11
### Fixed
- minor issues

## [2.1.0] - 2019-09-09
### Added
- support for driver version >3.2

## [2.0.2] - 2019-09-09
### Fixed
- battery status message when status is unavailable

## [2.0.1] - 2019-06-18
### Fixed
- minor manpage defects

## [2.0.0] - 2019-06-17
### Added
- `.desktop` file
- `-n` option for *not* saving battery thresholds for persistence
### Removed
- attempts to `chmod` files in `/etc/default/huawei-wmi/`
### Deprecated
- `-s` option ("save battery thresholds for persistence")

## [1.3.3] - 2019-06-10
### Fixed
- lack of manpage
- `.deb` did not depend on `libc6`

## [1.3.2] - 2019-06-07
### Fixed
- lack of `.deb` in release

## [1.3.1] - 2019-06-06
### Removed
- obsolete files in release

## [1.3.0] - 2019-06-06
### Added
- `-s` option for saving battery thresholds for persistence
### Changed
- outsorced udev rules to [qu1x/huawei-wmi](https://github.com/qu1x/huawei-wmi) project
### Fixed
- build process

## [1.2.5] - 2019-05-02
### Added
- `-wait` option for MateBook X
### Removed
- driver waits by default

## [1.2.4] - 2019-04-28
### Fixed
- battery protection status messages inconsistency

## [1.2.3] - 2019-04-28
### Added
- `-icon` option for custom icon
### Fixed
- didn't wait for driver to set thresholds
- minor glitches

## [1.2.2] - 2019-04-26
### Fixed
- icon was invisible on dark panel background

## [1.2.1] - 2019-04-26
### Fixed
- no version info was shown
- minor glitches

## [1.2.0] - 2019-04-25
### Added
- support for Huawei-WMI driver

## [1.1.0] - 2019-04-19
### Added
- Fn-Lock functionality
### Fixed
- minor glitches

## [1.0.1] - 2019-04-19
### Fixed
- icon not showing up

## 1.0.0 - 2019-04-19
*initial release*
### Added
- setting battery protection thresholds
- checks for scripts availability

[Unreleased]: https://github.com/nekr0z/matebook-applet/compare/3.1.0...HEAD
[3.1.0]: https://github.com/nekr0z/matebook-applet/compare/3.0.3...3.1.0
[3.0.3]: https://github.com/nekr0z/matebook-applet/compare/3.0.2...3.0.3
[3.0.2]: https://github.com/nekr0z/matebook-applet/compare/3.0.1...3.0.2
[3.0.1]: https://github.com/nekr0z/matebook-applet/compare/3.0.0...3.0.1
[3.0.0]: https://github.com/nekr0z/matebook-applet/compare/2.5.0...3.0.0
[2.5.0]: https://github.com/nekr0z/matebook-applet/compare/2.4.8...2.5.0
[2.4.8]: https://github.com/nekr0z/matebook-applet/compare/2.4.7...2.4.8
[2.4.7]: https://github.com/nekr0z/matebook-applet/compare/2.4.6...2.4.7
[2.4.6]: https://github.com/nekr0z/matebook-applet/compare/2.4.5...2.4.6
[2.4.5]: https://github.com/nekr0z/matebook-applet/compare/2.4.4...2.4.5
[2.4.4]: https://github.com/nekr0z/matebook-applet/compare/2.4.3...2.4.4
[2.4.3]: https://github.com/nekr0z/matebook-applet/compare/2.4.2...2.4.3
[2.4.2]: https://github.com/nekr0z/matebook-applet/compare/2.4.1...2.4.2
[2.4.1]: https://github.com/nekr0z/matebook-applet/compare/2.4.0...2.4.1
[2.4.0]: https://github.com/nekr0z/matebook-applet/compare/2.3.3...2.4.0

build.go
// Copyright (C) 2019 Evgeny Kuznetsov (evgeny@kuznetsov.md)
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

// +build ignore

package main

import (
	"archive/tar"
	"bufio"
	"bytes"
	"compress/gzip"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/nekr0z/changelog"
)

type packFile struct {
	src string
	dst string
	mod os.FileMode
}

type distFile struct {
	src string
	dst string
}

var (
	filename  string
	keyID     string = "FA32B7DDA1A3AC2C"
	packFiles        = []packFile{
		{src: "LICENSE", dst: "LICENSE", mod: 0644},
		{src: "README.md", dst: "README.md", mod: 0644},
		{src: "SOURCE.txt", dst: "SOURCE.txt", mod: 0644},
		{src: "matebook-applet.1", dst: "matebook-applet.1", mod: 0644},
		{src: "CHANGELOG.md", dst: "CHANGELOG.md", mod: 0644},
	}
	distFiles = []distFile{
		{src: "LICENSE", dst: "/usr/share/doc/matebook-applet/"},
		{src: "README.md", dst: "/usr/share/doc/matebook-applet/"},
		{src: "SOURCE.txt", dst: "/usr/share/doc/matebook-applet/"},
		{src: "matebook-applet.1", dst: "/usr/share/man/man1/"},
		{src: "matebook-applet.desktop", dst: "/usr/share/applications/"},
		{src: "assets/matebook-applet.png", dst: "/usr/share/icons/hicolor/512x512/apps/"},
	}
	debDeps = []string{
		"libayatana-appindicator3-1",
		"libc6",
		"libgtk-3-0 >= 3.10",
	}
	debRecs = []string{
		"huawei-wmi",
	}
)

func main() {
	sign := flag.Bool("s", false, "sign binary")
	tar := flag.Bool("t", false, "generate tar.gz")
	deb := flag.Bool("d", false, "build .deb")
	legacy := flag.Bool("l", false, "build against libappindicator")
	flag.Parse()
	version := getVersion()
	if *legacy {
		version = version + "-legacy"
	}
	btime := buildTime()
	if *tar {
		*sign = true
	}

	fmt.Printf("Building version %s\n", version)
	fmt.Println("Building as of", time.Unix(btime, 0))
	buildAssets(btime)
	buildBinary(version, btime)

	if *sign {
		signFile("matebook-applet", keyID)
	}

	if *tar {
		filename = "matebook-applet-amd64" + "-" + version
		buildTar()
		fmt.Println("archive", filename, "created")
	}

	if *deb {
		buildDeb(version)
	}
}

func buildBinary(version string, t int64) {
	cmdline := fmt.Sprintf("go build -buildmode=pie -trimpath -ldflags=\"-s -w -buildid= -X main.version=%s\"", version)
	if strings.HasSuffix(version, "-legacy") {
		cmdline = cmdline + " -tags legacyappindicator"
	}
	cmd := exec.Command("bash", "-c", cmdline)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stdout
	if err := cmd.Run(); err != nil {
		log.Fatalln("failed to build binary")
	}
	setFileTime("matebook-applet", t)
	packFiles = append(packFiles, packFile{"matebook-applet", "matebook-applet", 0755})
	distFiles = append(distFiles, distFile{"matebook-applet", "/usr/bin/"})
}

func buildAssets(t int64) {
	cmd := exec.Command("go", "run", "assets_generate.go")
	if err := cmd.Run(); err != nil {
		log.Fatalln("failed to rebuild assets")
	}
	setFileTime("assets.go", t)
}

func buildDeb(ver string) {
	maintainer := changelog.Maintainer{Name: "Evgeny Kuznetsov", Email: "evgeny@kuznetsov.md"}
	var clok bool
	fd, err := os.Open("CHANGELOG.md")
	if err != nil {
		fmt.Printf("error opening changelog: %s\n", err)
	} else {
		defer fd.Close()
		cl, err := changelog.ParseMd(fd)
		if err != nil {
			fmt.Printf("error parsing changelog: %s\n", err)
		}
		cmd := exec.Command("git", "tag", "-l", `--format=%(creatordate:iso)|%(refname:short)`)
		var bb bytes.Buffer
		out := bufio.NewWriter(&bb)
		cmd.Stdout = out
		if err := cmd.Run(); err != nil {
			fmt.Printf("failed to read tag times: %s", err)
		} else {
			out.Flush()
			scanner := bufio.NewScanner(&bb)
			for scanner.Scan() {
				line := scanner.Text()
				s := strings.Split(line, "|")
				if len(s) == 2 {
					d, err := time.Parse("2006-01-02 15:04:05 -0700", s[0])
					if err == nil {
						ver, err := changelog.ToVersion(strings.TrimPrefix(s[1], "v"))
						if err == nil {
							rel := cl[ver]
							rel.Date = d
							cl[ver] = rel
						}
					}
				}
			}
		}
		for v, rel := range cl {
			rel.Maintainer = maintainer
			cl[v] = rel
		}
		b, err := cl.Debian("matebook-applet")
		if err != nil {
			fmt.Printf("error converting changelog to Debian format: %s\n", err)
		}
		clDeb, err := os.Create("debian.changelog")
		if err != nil {
			fmt.Printf("error creating Debian changelog: %s\n", err)
		} else {
			defer clDeb.Close()
			_, err := clDeb.Write(b)
			if err != nil {
				fmt.Printf("error writing Debian changelog: %s\n", err)
			}
			clDeb.Sync()
			clok = true
		}
	}
	ver = strings.Replace(strings.TrimPrefix(ver, "v"), "-", "~", 1)
	args := []string{
		"-f",
		"-t", "deb",
		"-s", "dir",
		"-n", "matebook-applet",
		"-v", ver,
		"-m", fmt.Sprintf("%s <%s>", maintainer.Name, maintainer.Email),
		"--vendor", fmt.Sprintf("%s <%s>", maintainer.Name, maintainer.Email),
		"--category", "misc",
		"--description", "System tray applet for Huawei MateBook\nAllows one to control Huawei MateBook features,\nlike Fn-Lock and Battery Protection settings, via GUI.",
		"--url", "https://github.com/nekr0z/matebook-applet",
		"--license", "GPL-3",
		"--deb-priority", "optional",
	}
	if clok {
		args = append(args, "--deb-changelog", "debian.changelog")
	}
	for _, dep := range debDeps {
		args = append(args, "-d", dep)
	}
	for _, rec := range debRecs {
		args = append(args, "--deb-recommends", rec)
	}
	for _, file := range distFiles {
		arg := file.src + "=" + file.dst
		args = append(args, arg)
	}
	cmd := exec.Command("fpm", args...)
	if err := cmd.Run(); err != nil {
		fmt.Println(err)
		log.Fatalln("failed to build .deb")
	}
	fmt.Println(".deb package created")
}

func buildTar() {
	for i := range packFiles {
		packFiles[i].dst = filename + "/" + packFiles[i].dst
	}
	filename = filename + ".tar.gz"
	fd, err := os.Create(filename)
	if err != nil {
		log.Fatal(err)
	}
	gw, err := gzip.NewWriterLevel(fd, gzip.BestCompression)
	if err != nil {
		log.Fatal(err)
	}
	tw := tar.NewWriter(gw)
	for _, f := range packFiles {
		sf, err := os.Open(f.src)
		if err != nil {
			log.Fatal(err)
		}
		info, err := sf.Stat()
		if err != nil {
			log.Fatal(err)
		}
		h := &tar.Header{
			Name:    f.dst,
			Size:    info.Size(),
			Mode:    int64(f.mod),
			ModTime: info.ModTime(),
		}
		err = tw.WriteHeader(h)
		if err != nil {
			log.Fatal(err)
		}
		_, err = io.Copy(tw, sf)
		if err != nil {
			log.Fatal(err)
		}
		sf.Close()
	}
	err = tw.Close()
	if err != nil {
		log.Fatal(err)
	}
	err = gw.Close()
	if err != nil {
		log.Fatal(err)
	}
	err = fd.Close()
	if err != nil {
		log.Fatal(err)
	}
}

func setFileTime(f string, t int64) {
	cmd := exec.Command("touch", "-t", fmt.Sprint(time.Unix(t, 0).Format("200601021504.05")), f)
	if err := cmd.Run(); err != nil {
		log.Fatalln("failed to set time on", f)
	}
}

func signFile(f string, k string) {
	cmd := exec.Command("gpg", "--detach-sign", "--yes", "--passphrase", os.Getenv("GPG_PASSPHRASE"), "--pinentry-mode", "loopback", "-a", "-u", k, f)
	cmd.Stderr = os.Stdout
	cmd.Stdout = os.Stdout
	if err := cmd.Run(); err != nil {
		fmt.Println("signing", f, "failed")
		filename = filename + "-unsigned"
	} else {
		fmt.Println(f, "successfully signed with key", k)
		packFiles = append(packFiles, packFile{"matebook-applet.asc", "matebook-applet.asc", 0644})
	}
}

func getVersion() string {
	desc, err1 := getString("git", "describe", "--always", "--dirty")
	br, err2 := getString("git", "symbolic-ref", "--short", "-q", "HEAD")
	if err1 == nil && (br == "" || err2 == nil) {
		return parseVersion(desc, br)
	}
	return "unknown"
}

func parseVersion(desc, branch string) string {
	descRe := regexp.MustCompile(`^v?(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))??(?:\-(?P<after>0|[1-9]\d*)\-(?P<commit>g[0-9a-f]{5,15}))?(?:\-(?P<dirty>dirty))??$`)
	if !descRe.MatchString(desc) {
		return "unknown"
	}

	ver := descRe.ReplaceAllString(desc, "${major}.${minor}.${patch}")
	pre := descRe.ReplaceAllString(desc, "${prerelease}")
	after := descRe.ReplaceAllString(desc, "${after}")
	commit := descRe.ReplaceAllString(desc, "${commit}")
	dirty := descRe.ReplaceAllString(desc, "${dirty}")

	if pre != "" {
		ver = fmt.Sprintf("%s-%s", ver, pre)
	}

	if (after != "" || dirty != "") && branch != "master" && branch != "" {
		ver = fmt.Sprintf("%s-%s", ver, branch)
	}

	if after != "" {
		ver = fmt.Sprintf("%s-%s", ver, after)
	}

	if dirty != "" {
		ver = fmt.Sprintf("%s-%s", ver, "dirty")
	}

	if commit != "" {
		ver = fmt.Sprintf("%s+%s", ver, commit)
	}

	return ver
}

func getString(c string, a ...string) (string, error) {
	cmd := exec.Command(c, a...)
	b, err := cmd.CombinedOutput()
	return string(bytes.TrimSpace(b)), err
}

func buildTime() int64 {
	s, err := getString("git", "show", "-s", "--format=%ct")
	if err == nil {
		if i, e := strconv.ParseInt(s, 10, 64); e == nil {
			return i
		}
	}
	return time.Now().Unix()
}
